src/pointer.rs

```diff
@@ -845,6 +981,153 @@ const fn validate(value: &str) -> Result<&str, ParseError> {
+#[derive(Debug, PartialEq)]
+pub enum ParseError {
+    /// `Pointer` did not start with a backslash (`'/'`).
+    NoLeadingBackslash,
+
+    /// `Pointer` contained invalid encoding (e.g. `~` not followed by `0` or
+    /// `1`).
+    InvalidEncoding {
+        /// Offset of the partial pointer starting with the token that contained
+        /// the invalid encoding
+        offset: usize,
+        /// The source `InvalidEncodingError`
+        source: InvalidEncodingError,
+    },
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::NoLeadingBackslash { .. } => {
+                write!(
+                    f,
+                    "json pointer is malformed as it does not start with a backslash ('/')"
+                )
+            }
+            Self::InvalidEncoding { source, .. } => write!(f, "{source}"),
+        }
+    }
+}
+
+impl ParseError {
+    /// Returns `true` if this error is `NoLeadingBackslash`; otherwise returns
+    /// `false`.
+    pub fn is_no_leading_backslash(&self) -> bool {
+        matches!(self, Self::NoLeadingBackslash { .. })
+    }
+
+    /// Returns `true` if this error is `InvalidEncoding`; otherwise returns
+    /// `false`.
+    pub fn is_invalid_encoding(&self) -> bool {
+        matches!(self, Self::InvalidEncoding { .. })
+    }
+
+    /// Offset of the partial pointer starting with the token which caused the
+    /// error.
+    /// ```text
+    /// "/foo/invalid~tilde/invalid"
+    ///      ↑
+    ///      4
+    /// ```
+    /// ```
+    /// # use jsonptr::PointerBuf;
+    /// let err = PointerBuf::parse("/foo/invalid~tilde/invalid").unwrap_err();
+    /// assert_eq!(err.pointer_offset(), 4)
+    /// ```
+    pub fn pointer_offset(&self) -> usize {
+        match *self {
+            Self::NoLeadingBackslash { .. } => 0,
+            Self::InvalidEncoding { offset, .. } => offset,
+        }
+    }
+
+    /// Offset of the character index from within the first token of
+    /// [`Self::pointer_offset`])
+    /// ```text
+    /// "/foo/invalid~tilde/invalid"
+    ///              ↑
+    ///              8
+    /// ```
+    /// ```
+    /// # use jsonptr::PointerBuf;
+    /// let err = PointerBuf::parse("/foo/invalid~tilde/invalid").unwrap_err();
+    /// assert_eq!(err.source_offset(), 8)
+    /// ```
+    pub fn source_offset(&self) -> usize {
+        match self {
+            Self::NoLeadingBackslash { .. } => 0,
+            Self::InvalidEncoding { source, .. } => source.offset,
+        }
+    }
+
+    /// Offset of the first invalid encoding from within the pointer.
+    /// ```text
+    /// "/foo/invalid~tilde/invalid"
+    ///              ↑
+    ///             12
+    /// ```
+    /// ```
+    /// # use jsonptr::PointerBuf;
+    /// let err = PointerBuf::parse("/foo/invalid~tilde/invalid").unwrap_err();
+    /// assert_eq!(err.pointer_offset(), 4)
```

Comment:
Looks like this example is mismatched

Reply:
Wow, great catch. Damn you're great at this.